src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/HtmlFormattingPass.cs

```diff
@@ -125,6 +125,48 @@ private static List<TextChange> AdjustRazorIndentation(FormattingContext context
 
                 if (context.Indentations[i].StartsInCSharpContext)
                 {
+                    // Normally we don't do HTML things in C# contexts but there is one
+                    // edge case when including render fragments in a C# code block, eg:
+                    //
+                    // @code {
+                    //      void Foo()
+                    //      {
+                    //          Render(@<SurveyPrompt />);
+                    //      {
+                    // }
+                    //
+                    // This is popular in the bUnit unit testing library. The issue here is that
+                    // the HTML formatter sees ~~~~~<SurveyPrompt /> and puts a newline before
+                    // the tag, but obviously that breaks things.
+                    //
+                    // It's straight forward enough to just check for this situation and special case
+                    // it by removing the newline again.
+
+                    // There needs to be at least one more line, and the current line needs to end with
+                    // an @ sign, and have an open angle bracket at the start of the next line.
+                    if (sourceText.Lines.Count >= i + 1 &&
```

Comment:
Not really. Essentially since the HTML formatter is breaking this, and we don't control it, I wanted to fix it as close as possible to the break, so the rest of our formatting engine can continue as normal. I actually even considered filtering out the edit itself from the formatter results, but that had a little bit too little context to be reliable.

I _could_ move this to somewhere that we have a syntax tree etc. but by then other formatting operations would have occurred.. eg HTML formats this to two lines, then the C# formatter sees the `<` and treats it as a greater-than and adds a space after it, etc. etc. so whilst there would be more information available, there would also be more to fix up. I think its nicer that this is just removing a single newline.

Reply:
Ahhh I see. Kinda curious, doesn't the old formatting cause a Razor compiler error?

Reply:
Thats a very good point.. it certainly has C# compiler errors, but no Razor compiler errors.
<img width="280" alt="image" src="https://user-images.githubusercontent.com/754264/144932414-220adf63-a8ba-4d49-9545-da9f0f7d5cf8.png">

Converting the `ValidateComponents` check I put in from being debug-only to be a validation pass would help here, but at the same time it would throw out all formatting operations and hide this issue, and I prefer that we found this and worked around it.

Reply:
Deffffinitely prefer your approach too. Interesting that it's treated as C#. I guess the Razor compiler is a little bit smart where `@`'s can exist